src/protagonist/API/Features/Image/Ingest/DeliveryChannelProcessor.cs

```diff
@@ -0,0 +1,209 @@
+
+                assetImageDeliveryChannels.Add(new ImageDeliveryChannel
+                {
+                    ImageId = assetId,
+                    DeliveryChannelPolicyId = deliveryChannelPolicy.Id,
+                    Channel = deliveryChannel.Channel
+                });
+                changeMade = true;
+            }
+            else
+            {
+                // There is already a IDC for this thing - has the policy changed?
+                if (currentChannel.DeliveryChannelPolicyId != deliveryChannelPolicy.Id)
+                {
+                    logger.LogTrace(
+                        "Asset {AssetId} already has deliveryChannel {DeliveryChannel}, but policy changed from {OldPolicyName} to Asset {NewPolicyName}",
+                        assetId, deliveryChannel.Channel, currentChannel.DeliveryChannelPolicy.Name,
+                        deliveryChannelPolicy.Name);
+                    currentChannel.DeliveryChannelPolicy = deliveryChannelPolicy;
+                    changeMade = true;
+                }
+            }
+        }
+
+        if (isUpdate)
+        {
+            // Remove any that are no longer part of the payload
+            foreach (var deletedChannel in assetImageDeliveryChannels.Where(idc =>
+                         !handledChannels.Contains(idc.Channel)))
+            {
+                logger.LogTrace("Removing deliveryChannel {DeliveryChannel}, from Asset {AssetId}",
+                    deletedChannel.Channel, assetId);
+                assetImageDeliveryChannels.Remove(deletedChannel);
+                changeMade = true;
+            }
+        }
+            
+        return changeMade;
+    }
+
+    private async Task<DeliveryChannelPolicy> GetDeliveryChannelPolicy(Asset asset, DeliveryChannelsBeforeProcessing deliveryChannel)
+    {
+        DeliveryChannelPolicy deliveryChannelPolicy;
+        if (deliveryChannel.Policy.IsNullOrEmpty())
+        {
+            deliveryChannelPolicy = await defaultDeliveryChannelRepository.MatchDeliveryChannelPolicyForChannel(
+                asset.MediaType!, asset.Space, asset.Customer, deliveryChannel.Channel);
+        }
+        else
+        {
+            deliveryChannelPolicy = await deliveryChannelPolicyRepository.RetrieveDeliveryChannelPolicy(
+                asset.Customer,
+                deliveryChannel.Channel,
+                deliveryChannel.Policy);
+        }
+
+        return deliveryChannelPolicy;
+    }
+
+    private void AddExplicitNoneChannel(Asset asset)
+    {
+        logger.LogTrace("assigning 'none' channel for asset {AssetId}", asset.Id);
+        var deliveryChannelPolicy = deliveryChannelPolicyRepository.RetrieveDeliveryChannelPolicy(asset.Customer,
+            AssetDeliveryChannels.None, None);
+
+        // "none" channel can only exist on it's own so remove any others that may be there already prior to adding
+        asset.ImageDeliveryChannels.Clear();

```

Comment:
does this mean that if you had `["file","none"]`, the `file` channel would be dropped silently?

Reply:
No, this method can only be entered if there is a single delivery-channel and it's `"none"` (validation will catch this higher up too) so in this case `asset.ImageDeliveryChannels.Clear()` is just removing any existing existing IDC records from database (e.g. if this was `["file", "iiif-img", "thumbs"]` it'll remove them and them add the `"none"` record).